src/LandIce/evaluators/LandIce_BasalMeltRate_Def.hpp

```diff
@@ -107,55 +120,173 @@ postRegistrationSetup(typename Traits::SetupData d, PHX::FieldManager<Traits>& f
+          basalVertVelocity(cell,side,node) =  - melting /(1 - rho_w/rho_i*std::min(phi(sideSet_idx,node),0.5)) -  basal_water_flux;
+        }
+      }
+    } else {
+      for (int sideSet_idx = 0; sideSet_idx < sideSet.size; ++sideSet_idx)
       {
-        //always in presence of water on shelves (assuming that beta==0 <==> on shelves)
-        bool isThereWaterHere = isThereWater || (beta(cell,side,node) == 0.0);
-        ScalarT diffEnthalpy = Enthalpy(cell,side,node) - EnthalpyHs(cell,side,node);
-        ScalarT basal_reg_scale = (diffEnthalpy > 0 || !isThereWaterHere) ?  ScalarT(0.5 + 0.5*tanh(basal_reg_coeff * diffEnthalpy)) :
-                                                                             ScalarT(0.5 + 0.5* basal_reg_coeff * diffEnthalpy);
-                                                                     //    ScalarT(0.5 + 0.5* (0.5-0.5*std::pow(1-basal_reg_coeff * diffEnthalpy,2)));
-
-        //mstar, [W m^{-2}] = [Pa m s^{-1}]: basal latent heat in temperate ice
-        ScalarT mstar = geoFluxHeat(cell,side,node);
-        for (int dim = 0; dim < vecDimFO; dim++)
-          mstar += 1000./scyr * beta(cell,side,node) * velocity(cell,side,node,dim) * velocity(cell,side,node,dim);
-
-        double dTdz_melting = beta_p * rho_i * g;
-        mstar += 1e-3* k_i * dTdz_melting;
-
-        enthalpyBasalFlux(cell,side,node) =  (basal_reg_scale-1) *mstar + 1e-3*k_i*dTdz_melting;
-
-        ScalarT basal_water_flux = scyr * k_0 * (rho_w - rho_i) * g / eta_w * pow(phi(cell,side,node),alpha_om); //[m yr^{-1}]
-        ScalarT melting = scyr * basal_reg_scale * mstar / (L*rho_i); //[m yr^{-1}]
-        basalVertVelocity(cell,side,node) =  - melting /(1 - rho_w/rho_i*std::min(phi(cell,side,node),0.5)) -  basal_water_flux;
+        // Get the local data of side and cell
+        const int cell = sideSet.elem_LID(sideSet_idx);
+        const int side = sideSet.side_local_id(sideSet_idx);
+
+        for (int node = 0; node < dim; ++node)
+        {
+          //always in presence of water on shelves (assuming that beta==0 <==> on shelves)
+          bool isThereWaterHere = isThereWater || (beta(cell,side,node) == 0.0);
+          ScalarT diffEnthalpy = Enthalpy(cell,side,node) - EnthalpyHs(cell,side,node);
+          ScalarT basal_reg_scale = (diffEnthalpy > 0 || !isThereWaterHere) ?  ScalarT(0.5 + 0.5*tanh(basal_reg_coeff * diffEnthalpy)) :
+                                                                              ScalarT(0.5 + 0.5* basal_reg_coeff * diffEnthalpy);
+                                                                      //    ScalarT(0.5 + 0.5* (0.5-0.5*std::pow(1-basal_reg_coeff * diffEnthalpy,2)));
+
+          //mstar, [W m^{-2}] = [Pa m s^{-1}]: basal latent heat in temperate ice
+          ScalarT mstar = geoFluxHeat(cell,side,node);
+          for (int dim = 0; dim < vecDimFO; dim++)
+            mstar += 1000./scyr * beta(cell,side,node) * velocity(cell,side,node,dim) * velocity(cell,side,node,dim);
```

Comment:
1000./scyr could be pre-computed. Alternatively, you could make `scyr` a `static constexpr double` (and init it in the class), so that `1000.0/scyr` is computed at compile time. You might even consider pulling scyr outside the class, and just declare it locally in the fcn as a `constexpr double`, since it is not really part of the "state" of the object.